### DISCUSSION TITLE
Which Postgres user does the Supabase client use?

### DISCUSSION THREAD

**Discussion by liaujianjie** [supabase/supabase#2027] [Questions] [answered]
I wish to implement some column-level security (e.g. preventing updates of column after creation) using the Postgres `REVOKE` SQL command.

As I understand it, there are 6 Postgres users created by default:

```
supabase_storage_admin	
supabase_auth_admin	
supabase_admin
postgres
pgbouncer	
authenticator	
```

On which Postgres user should I be applying such queries?

Another related question: which Postgres user is the dashboard using? I want to still have the ability to modify the values on the dashboard whilst disallowing my users from modifying the columns on their client.

**Comment by soedirgo** [comment, marked as answer]
Hey @liaujianjie,

> Which Postgres user does the Supabase client use?

This depends on which API key you use—`anon` role for the `anon`/`public` key, and `service_role` for the `service_role`/`secret` key. You can mint your own key for the appropriate role using the JWT secret.

> which Postgres user is the dashboard using?

The dashboard uses the `supabase_admin` role.

Guessing from your requirements, you probably want to use the `anon` key for your users.

**Reply by liaujianjie** [reply]
Revoking on both `anon` and `authenticated` roles still doesn't prevent the JS client from updating the columns.

```sql
REVOKE UPDATE ("read_only_column") ON test_table FROM anon, authenticated;
```

```js
const updateResponse = await supabase
  .from("test_table")
  .update({ read_only_column: "mmuahahaha evil change" });
console.log(updateResponse.statusText); // prints: OK
```

**Reply by liaujianjie** [reply]
In fact, revoking `PUBLIC` also doesn't prevent the JS client from updating the columns.

```sql
REVOKE UPDATE ("read_only_column") ON test_table FROM PUBLIC;
```

**Reply by soedirgo** [reply]
That may not be enough to revoke access from `anon`, you'll also need to revoke privileges from roles `anon` is a member of as well as privileges granted by other roles. More info in the [Postgres docs](https://www.postgresql.org/docs/current/sql-revoke.html).

You can try running the following SQL to see if it still has privileges (adapted from [postgres-meta](https://github.com/supabase/postgres-meta)):

<details>
<summary>Click to expand</summary>

```sql
-- Adapted from information_schema.role_table_grants

SELECT
  c.oid :: int8 AS table_id,
  u_grantor.rolname AS grantor,
  grantee.rolname AS grantee,
  nc.nspname AS schema,
  c.relname AS table_name,
  c.prtype AS privilege_type,
  CASE
    WHEN pg_has_role(grantee.oid, c.relowner, 'USAGE')
    OR c.grantable THEN TRUE
    ELSE FALSE
  END AS is_grantable,
  CASE
    WHEN c.prtype = 'SELECT' THEN TRUE
    ELSE FALSE
  END AS with_hierarchy
FROM
  (
    SELECT
      pg_class.oid,
      pg_class.relname,
      pg_class.relnamespace,
      pg_class.relkind,
      pg_class.relowner,
      (
        aclexplode(
          COALESCE(
            pg_class.relacl,
            acldefault('r', pg_class.relowner)
          )
        )
      ).grantor AS grantor,
      (
        aclexplode(
          COALESCE(
            pg_class.relacl,
            acldefault('r', pg_class.relowner)
          )
        )
      ).grantee AS grantee,
      (
        aclexplode(
          COALESCE(
            pg_class.relacl,
            acldefault('r', pg_class.relowner)
          )
        )
      ).privilege_type AS privilege_type,
      (
        aclexplode(
          COALESCE(
            pg_class.relacl,
            acldefault('r', pg_class.relowner)
          )
        )
      ).is_grantable AS is_grantable
    FROM
      pg_class
  ) c(
    oid,
    relname,
    relnamespace,
    relkind,
    relowner,
    grantor,
    grantee,
    prtype,
    grantable
  ),
  pg_namespace nc,
  pg_authid u_grantor,
  (
    SELECT
      pg_authid.oid,
      pg_authid.rolname
    FROM
      pg_authid
    UNION ALL
    SELECT
      0 :: oid AS oid,
      'PUBLIC'
  ) grantee(oid, rolname)
WHERE
  c.relnamespace = nc.oid
  AND (c.relkind IN ('r', 'v', 'f', 'p'))
  AND c.grantee = grantee.oid
  AND c.grantor = u_grantor.oid
  AND (
    c.prtype IN (
      'INSERT',
      'SELECT',
      'UPDATE',
      'DELETE',
      'TRUNCATE',
      'REFERENCES',
      'TRIGGER'
    )
  )
  AND (
    pg_has_role(u_grantor.oid, 'USAGE')
    OR pg_has_role(grantee.oid, 'USAGE')
  )
```
</details>

If you created the table from the dashboard it should be enough to revoke it from `supabase_admin` on `anon` (or at least it works on my end).

**Reply by soedirgo** [reply]
I think I got it now, the privilege for the table overrides the privilege for the column, so instead you want to revoke all privileges on the table from `anon`, and then grant privileges for individual columns on `anon`. A bit tedious, that...

**Reply by liaujianjie** [reply]
Thanks you so much for looking into it, I'll give it a try tomorrow morning and let you know if it worked. :-)

**Reply by liaujianjie** [reply]
Hey @soedirgo, I can confirm that revoking the privileges on a table does indeed remove the privileges like you mentioned. Just one thing: It seems that authenticated users use the `authenticated` Postgres role. As such, I revoked the privileges on the table for both `anon` and `authenticated`.

**Reply by liaujianjie** [reply]
Thanks for your help!

**Reply by soedirgo** [reply]
No probs—glad you sorted it out :)
